### DISCUSSION TITLE
Ability to turn off directory structure affecting component names?

### DISCUSSION THREAD

**Discussion by jerryjappinen** [nuxt/nuxt#17167] [General]
I'm looking at starting to build/migrate bigger sites to Nuxt, and I find the component names being generated based on the directory structure a bit problematic.

In the docs, there's this simple example:

```
| components/
--| base/
----| foo/
------| Button.vue
```

Which gives you `<BaseFooButton />`. But when you start having more components, you're going to want to organise them into subfolders:

```
| components/
--| profile/
----| ProfileCoverImage.vue
----| ProfileIconButtons.vue
----| ProfileIntroText.vue
----| ProfileStats.vue
--| transitions/
----| ContentTransition.vue
----| ProfileIconButtonsTransition.vue
----| ProfileIntroTextTransition.vue
----| SlowContentTransition.vue
--| layout/
----| SmallSubtitle.vue
----| Subtitle.vue
...
```

I'd like the file name to always match the component's name in use. This way I can easily search and pinpoint the files, see the matching file names in my editor and so on. Even [the documentation says](https://v3.nuxtjs.org/guide/directory-structure/components#component-names) that the filenames should match.

But it seems like there is no way to organise the components in the `components/` directory anymore without breaking this.

You can probably learn to game the new system to some extent. But here's the best I could come up with just for the use case above:

```
| components/
--| Profile/
----| CoverImage.vue
----| IconButtons.vue
----| IntroText.vue
----| Stats.vue
--| Transition/
----| Content.vue
----| ProfileIconButtons.vue
----| ProfileIntroText.vue
----| SlowContent.vue
--| SmallSubtitle.vue
--| Subtitle.vue
...
```

This has a few issues:

- File names no longer match component names (so your git sidebar on VS Code will be more confusing)
- Forces unnatural component naming order: (`<ContentTransition />` has to become `<TransitionContent />` just so I can organise all the transition components into one folder)
- When I search for a `IconButtons`, what kind of file listing will my editor show? I can no longer tell that `ProfileIconButtons.vue` is actually a transition, and not in the profile. I want all the transitions in one place, but this will no longer really be possible. The file name no longer has the full context.
- Have to accept unnatural and long component names (like `<LayoutSubtitle />`) or lose categorisation

The new system doesn't let me organise the files the way I find the most natural. This is a simple example, but a real non-trivial app or web site will easily have a hundred components, or multiple hundreds. I find this is very confusing.

---

Is there a way to change the component name generation so that the directory structure doesn't change the auto-imported component name? I'd much rather have the file name always represent the component name, and take a build error in case there are naming conflicts, and I had zero need for this change from Nuxt 2.

PS I might also have naming conflicts from modules and I'd expect an error message from those too. They should probably be prefixed by the module author, but it can happen.

**Comment by tylerforesthauser** [comment]
There are options available when adjusting your config for components. 

```js
export default defineNuxtConfig({
  components: [
    '~/components',
    { path: '~/components/nested', pathPrefix: false },
  ],
})
```
Hope that helps. 

[Reference](https://github.com/nuxt/framework/blob/main/packages/schema/src/types/components.ts#L23)

**Reply by jerryjappinen** [reply]
Thanks. I wouldn't want to have to do this for every folder separately, but I'll play around with this.

**Reply by tylerforesthauser** [reply]
Look into the `pattern` option, which accepts [glob patterns](https://github.com/isaacs/node-glob#glob-primer).
